prov/mrail/src/mrail_domain.c

```diff
@@ -0,0 +1,176 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+static int mrail_domain_close(fid_t fid)
+{
+	struct mrail_domain *mrail_domain =
+		container_of(fid, struct mrail_domain, util_domain.domain_fid.fid);
+	int ret, retv = 0;
+
+	if (mrail_domain->domains) {
+		ret = mrail_close_fids((struct fid **)mrail_domain->domains,
+				       mrail_domain->num_domains);
+		if (ret)
+			retv = ret;
+		free(mrail_domain->domains);
+	}
+
+	ret = ofi_domain_close(&mrail_domain->util_domain);
+	if (ret)
+		retv = ret;
+
+	free(mrail_domain);
+	return retv;
+}
+
+//static int mrail_mr_close(fid_t fid)
+//{
+//
+//	return -FI_ENOSYS;
+//}
+
+static int mrail_mr_reg(struct fid *domain_fid, const void *buf, size_t len,
+			 uint64_t access, uint64_t offset, uint64_t requested_key,
+			 uint64_t flags, struct fid_mr **mr, void *context)
+{
+
+	return -FI_ENOSYS;
+}
+
+//static struct fi_ops mrail_mr_ops = {
+//	.size = sizeof(struct fi_ops),
+//	.close = mrail_mr_close,
+//	.bind = fi_no_bind,
+//	.control = fi_no_control,
+//	.ops_open = fi_no_ops_open,
+//};
+
+static struct fi_ops_mr mrail_domain_mr_ops = {
+	.size = sizeof(struct fi_ops_mr),
+	.reg = mrail_mr_reg,
+	.regv = fi_no_mr_regv,
+	.regattr = fi_no_mr_regattr,
+};
+
+static struct fi_ops mrail_domain_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = mrail_domain_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_domain mrail_domain_ops = {
+	.size = sizeof(struct fi_ops_domain),
+	.av_open = fi_no_av_open,
+	.cq_open = mrail_cq_open,
+	.endpoint = mrail_ep_open,
+	.scalable_ep = fi_no_scalable_ep,
+	.cntr_open = fi_no_cntr_open,
+	.poll_open = fi_no_poll_open,
+	.stx_ctx = fi_no_stx_context,
+	.srx_ctx = fi_no_srx_context,
+	.query_atomic = fi_no_query_atomic,
+};
+
+int mrail_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		      struct fid_domain **domain, void *context)
+{
+	struct mrail_fabric *mrail_fabric =
+		container_of(fabric, struct mrail_fabric, util_fabric.fabric_fid);
+	struct mrail_domain *mrail_domain;
+	struct fi_info *fi, *core_info;
+	size_t i;
+	int ret;
+
+	mrail_domain = calloc(1, sizeof(*mrail_domain));
+	if (!mrail_domain)
+		return -FI_ENOMEM;
+
+	ret = ofi_domain_init(fabric, info, &mrail_domain->util_domain, context);
+	if (ret) {
+		free(mrail_domain);
+		return ret;
+	}
+
+	if (!mrail_fabric->fabrics) {
+		mrail_info_foreach(i, fi, info)
+			mrail_fabric->num_fabrics++;
+
+		mrail_fabric->fabrics = calloc(mrail_fabric->num_fabrics,
+					       sizeof(*mrail_fabric->fabrics));
+		if (!mrail_fabric->fabrics) {
+			ret = -FI_ENOMEM;
+			goto err1;
+		}
+	}
```

Comment:
I would have expected that the needed fabrics should all have been opened by now.  We just need to find the right one.